### DISCUSSION TITLE
Help in implementing data fetcher for a federated subgraph

### DISCUSSION THREAD

**Discussion by shamin2021** [graph-quilt/graphql-orchestrator-java#192] [Q&A]
I have the following two schemas 
Schema 1
type Query {
  getFoo: Foo
}

type Foo @key(fields: "id"){
    id: String
    fooName: String
}

Schema 2
type Query {
  getBar(barId: String): Bar
}

type Bar @key(fields: "id") {
    id: String
    barName: String
}

type Foo @extends @key(fields: "id") {
    id: String @external
    bar: Bar
}

and the service providers as 

Bar 

package org.example;

class ProductNameService implements ServiceProvider {

    String filePath1 = "src/main/resources/Product.graphql";

    public String schema;
    {
        try {
            schema = readFileAsString(filePath1);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public String getNameSpace() { return "PERSON_NAME"; }

    @Override
    public ServiceType getSeviceType() {
        return ServiceType.FEDERATION_SUBGRAPH;
    }

    @Override
    public Map<String, String> sdlFiles() { return ImmutableMap.of("schema.graphqls", schema); }

    @Override
    public CompletableFuture<Map<String, Object>> query(final ExecutionInput executionInput,
                                                        final GraphQLContext context) {
        return null;
    }

    public static String readFileAsString(String filePath) throws Exception {
        Path path = Paths.get(filePath);
        byte[] bytes = Files.readAllBytes(path);
        String content = new String(bytes, StandardCharsets.UTF_8);
        return content;
    }
}

and 

Food service 
package org.example;

import com.google.common.collect.ImmutableMap;
import com.intuit.graphql.orchestrator.ServiceProvider;
import graphql.ExecutionInput;
import graphql.GraphQLContext;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.CompletableFuture;

class ProductReviewService implements ServiceProvider {

    String filePath1 = "src/main/resources/Reviews.graphql";
    String filePath2 = "src/main/resources/resolverSchema.graphql";

    public  final String schema;
    public  final String resolverSchema;

    {
        try {
            schema = readFileAsString(filePath1);
            resolverSchema = readFileAsString(filePath2);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public String getNameSpace() { return "PERSON_ADDRESS";}

    @Override
    public Map<String, String> sdlFiles() { return ImmutableMap.of("schema.graphqls", schema, "resolver.graphqls", resolverSchema);}

    @Override
    public ServiceType getSeviceType() {
        return ServiceType.FEDERATION_SUBGRAPH;
    }

    @Override
    public CompletableFuture<Map<String, Object>> query (final ExecutionInput executionInput, final GraphQLContext context) {
        return null;

    }

    public static String readFileAsString(String filePath) throws Exception {
        Path path = Paths.get(filePath);
        byte[] bytes = Files.readAllBytes(path);
        String content = new String(bytes, StandardCharsets.UTF_8);
        return content;
    }

}

i have the main class as 

package org.example;
public class Test {

    public static void main(String[] args) throws Exception {

        // create a runtimeGraph by stitching service providers
        RuntimeGraph runtimeGraph = SchemaStitcher.newBuilder()
                .service(new ProductNameService())
                .service(new ProductReviewService())
                .build()
                .stitchGraph();

        String schema = new SchemaPrinter().print(runtimeGraph.getExecutableSchema());

        // pass the runtime graph to GraphQLOrchestrator
        GraphQLOrchestrator graphQLOrchestrator = GraphQLOrchestrator.newOrchestrator()
                .runtimeGraph(runtimeGraph).build();

        // Print SDL of supergraph
        String grastrator = new SchemaPrinter().print(graphQLOrchestrator.getSchema());
        System.out.println(grastrator);

        //Execute the query
        CompletableFuture<ExecutionResult> execute = graphQLOrchestrator
                .execute(
                        ExecutionInput.newExecutionInput()
                                .query( "query { getBar (barId: \"1\"){id} }")
                                .build()
                );

        //get execution result and print
        ExecutionResult executionResult = execute.get();
        System.out.println("--------------------------------------");
        System.out.println(executionResult);
    }

}

but when executed i get the following error please hep me resolve this 
ExecutionResultImpl{errors=[ExceptionWhileDataFetching{path=[getBar], exception=java.lang.IllegalAccessError: class com.intuit.graphql.orchestrator.common.ArgumentValueResolver tried to access private method 'void graphql.execution.ValuesResolver.<init>()' (com.intuit.graphql.orchestrator.common.ArgumentValueResolver and graphql.execution.ValuesResolver are in unnamed module of loader 'app'), locations=[SourceLocation{line=1, column=9}]}], data={getBar=null}, dataPresent=true, extensions=null}

**Comment by ashpak-shaikh** [comment]
you have to implement the `query` method in both the product name and review service and that needs to return the data against the query. When you execute the query against the orchestrator, it will make queries against the service provider. Can u post an example of this in a git repo so I can take a look at it? 

Also, you can try the gateway locally that has sample subgraphs in the local registry folder. 

https://github.com/graph-quilt/graphql-gateway-java

**Reply by shamin2021** [reply]
https://github.com/shamin2021/graphql-java.git 
this is the git repo with the code in the [Graphql-schemaregistry] branch 

I am new to graphql it would be great if you could give me some pointers therefore 
thankyou very much
